src/anyp/Uri.cc

```diff
@@ -937,55 +947,62 @@ AnyP::Uri::Uri(AnyP::UriScheme const &aScheme) :
 }
 
 // TODO: fix code duplication with AnyP::Uri::parse()
-char *
-AnyP::Uri::cleanup(const char *uri)
+SBuf
+AnyP::Uri::Cleanup(SBuf &uri)
 {
+    // 'whitespace' for chop and strip are different
+    static const CharacterSet wspChop = (CharacterSet::WSP +
+                                         CharacterSet::LF +
+                                         CharacterSet::CR).rename("wsp-chop");
+    static const CharacterSet wspStrip = (CharacterSet("VT,FF","\v\f") +
+                                          wspChop).rename("wsp-strip");
+
     int flags = 0;
-    char *cleanedUri = nullptr;
+    SBuf cleanedUri;
     switch (Config.uri_whitespace) {
     case URI_WHITESPACE_ALLOW:
         flags |= RFC1738_ESCAPE_NOSPACE;
     // fall through to next case
     case URI_WHITESPACE_ENCODE:
         flags |= RFC1738_ESCAPE_UNESCAPED;
-        cleanedUri = xstrndup(rfc1738_do_escape(uri, flags), MAX_URL);
+        cleanedUri = SBuf(rfc1738_do_escape(uri.c_str(), flags));
         break;
 
     case URI_WHITESPACE_CHOP: {
         flags |= RFC1738_ESCAPE_UNESCAPED;
-        const auto pos = strcspn(uri, w_space);
-        char *choppedUri = nullptr;
-        if (pos < strlen(uri))
-            choppedUri = xstrndup(uri, pos + 1);
-        cleanedUri = xstrndup(rfc1738_do_escape(choppedUri ? choppedUri : uri, flags), MAX_URL);
-        cleanedUri[pos] = '\0';
-        xfree(choppedUri);
+        const auto pos = uri.findFirstOf(wspChop);
+        SBuf choppedUri = uri.substr(0, pos);
+        cleanedUri = SBuf(rfc1738_do_escape(choppedUri.c_str(), flags));
     }
     break;
 
     case URI_WHITESPACE_DENY:
     case URI_WHITESPACE_STRIP:
     default: {
         // TODO: avoid duplication with urlParse()
-        const char *t;
-        char *tmp_uri = static_cast<char*>(xmalloc(strlen(uri) + 1));
-        char *q = tmp_uri;
-        t = uri;
-        while (*t) {
-            if (!xisspace(*t)) {
-                *q = *t;
-                ++q;
+        const auto shiftFrom = uri.findFirstOf(wspStrip);
+        if (shiftFrom == SBuf::npos) {
+            cleanedUri = uri;
+            break; // nothing to do
+        }
+        char *tmp_uri = SBufToCstring(uri); // allocate and fill tmp_uri from uri
```

Comment:
Please add an `RFC1738_SKIP_SPACE` or a similar rfc1738_do_escape() flag and refactor this code to use that flag instead of this allocation/deallocation and the shifting loop. One allocation and loop is enough and rfc1738_do_escape() already has both.

To keep things fast and simple, the adjusted rfc1738_do_escape() function can continue to use xisspace() for bad character detection (until that terrible C code is rewritten to use SBuf; the latter rewrite is outside this PR scope).

Reply:
Adding more if-statements or switch cases to the URI-encoder would impact a lot of other uses of that coder. Especially if the addition had any bugs. This PR is intentionally restricted to the AnyP::Uri code.

There is also already a definition of "SPACE" in that encoder as just character 0x20 (SP). For this cleanup stripping any of the wspStrip characters are removed from the output string.

I have since found that isspace() changes its definition of whitespace character depending on locale. So am reverting the use of wspStrip here.

Reply:
> Adding more if-statements or switch cases to the URI-encoder would impact a lot of other uses of that coder

My suggestion has no impact on other callers because the added code will be conditional on the new flag that no caller will use that flag (by default).

> There is also already a definition of "SPACE" in that encoder as just character 0x20 (SP).

That fact is irrelevant. My suggestion still works. If you do not like RFC1738_SKIP_SPACE name for the new flag (e.g., because the "space" in that flag is different from the "space" in the existing rfc1738_do_escape() code), then use a different/better name. The spelling of that flag is of secondary importance, and I am sure we can find some name that works well enough.

> I have since found that isspace() changes its definition of whitespace character depending on locale. So am reverting this block of changes.

The current changes are similarly flawed so this change request remains unaddressed. AFAICT, the original suggestion is still valid.

However, there is now another way to address the same concern: You can add a constant SBuf iterator to remove that extra SBuf copy/c_str() call. Please let me know if you prefer that alternative solution and need help with it.

The original suggestion is better because it eliminates both extra memory allocations and duplicated looping. The alternative only eliminates extra memory allocations. However, the alternative is acceptable because it does not make things worse, is generally useful, and is simpler.

Reply:
This is added in 3effc15f5